src/Temporalio.Extensions.DiagnosticSource/CustomMetricMeter.cs

```diff
@@ -0,0 +1,188 @@
+
+        /// <inheritdoc />
+        public ICustomMetricGauge<T> CreateGauge<T>(
+            string name, string? unit, string? description)
+            where T : struct
+        {
+            var gauge = new CustomMetricGauge<T>(name);
+            Meter.CreateObservableGauge(name, gauge.Snapshot, unit, description);
+            return gauge;
+        }
+
+        /// <inheritdoc />
+        public object CreateTags(
+            object? appendFrom, IReadOnlyCollection<KeyValuePair<string, object>> tags) =>
+            new Tags(tags, appendFrom);
+
+        private sealed class CustomMetricCounter<T> : ICustomMetricCounter<T>
+            where T : struct
+        {
+            private readonly Counter<T> underlying;
+
+            internal CustomMetricCounter(Counter<T> underlying) => this.underlying = underlying;
+
+            public void Add(T value, object tags) =>
+                underlying.Add(value, ((Tags)tags).TagList);
+        }
+
+        private sealed class CustomMetricHistogram<T> : ICustomMetricHistogram<T>
+            where T : struct
+        {
+            private readonly Histogram<T> underlying;
+
+            internal CustomMetricHistogram(Histogram<T> underlying) => this.underlying = underlying;
+
+            public void Record(T value, object tags) =>
+                underlying.Record(value, ((Tags)tags).TagList);
+        }
+
+#pragma warning disable CA1001 // We are disposing the lock on destruction since this can't be disposable
+        private sealed class CustomMetricGauge<T> : ICustomMetricGauge<T>
+#pragma warning restore CA1001
+            where T : struct
+        {
+            private readonly string name;
+            // We must dedupe based on tag set. We considered several designs including regular
+            // dictionary, blocking collection or concurrent queue w/ read time dedupe, and others
+            // and it was clearest to use a concurrent dictionary for collection but lock it while
+            // collecting metrics. Iterating a concurrent dictionary does not guarantee duplicates
+            // will be prevented, so we must lock it while iterating during collection.
+            private readonly ConcurrentDictionary<Tags, Measurement<T>> pending = new();
+            private readonly ReaderWriterLockSlim pendingLock = new();
+
+            internal CustomMetricGauge(string name) => this.name = name;
+
+            ~CustomMetricGauge() => pendingLock.Dispose();
+
+            public void Set(T value, object tags)
+            {
+                // We need to support an extreme max here just to prevent unbounded memory growth.
+                // We do not want to require a logger for this, so at some extreme value we will
+                // just write to stderr and drop.
+                if (pending.Count > 50000)
+                {
+                    Console.Error.WriteLine($"Dropping gauge metric {name} since cardinality has grown over 50k");
```

Comment:
I understand this is a very rare case, but it seems dangerous as it currently stand. If ever that `pending.Count > 50000` condition is reached (eg. user got creative with their tags, or just didn't understand that adding "workflow id" as a tag isn't a good idea), then console will get flooded with that error message, possibly bringing down the log target... Can't we just empty `pending`, and move on?

Reply:
Silently emptying without telling a user is arguably more dangerous. Not only are you discarding 50k worth of valuable information but you aren't even telling them you did it. We can't reasonably weave a logger through here and we decided, same on the core side that will also spam on dropped metric when buffer is full in the buffer impl, that we can come back and debounce logs or open up more options if it's ever needed.